pkg/partmgr/partmgr.go

```diff
@@ -18,123 +18,201 @@ under the Apache 2.0 license is conditioned upon your compliance with
+		manager:           pmgr,
+		path:              path,
+		startTime:         startTime,
+		partitionInterval: partitionInterval,
+		chunkInterval:     chunkInterval,
+		prefix:            "",
+		retentionDays:     pmgr.cfg.PartitionSchemaInfo.SampleRetention,
+		rollupTime:        rollupTime,
 	}
 
-	aggrType, _ := aggregate.AggrsFromString(pmgr.cfg.DefaultRollups) // TODO: error check & load part data from schema object
+	aggrType, _ := aggregate.AggrsFromString(pmgr.cfg.PartitionSchemaInfo.Aggregators)
 	newPart.defaultRollups = aggrType
-	if pmgr.cfg.RollupMin != 0 {
-		newPart.rollupBuckets = pmgr.cfg.DaysPerObj * 24 * 60 / pmgr.cfg.RollupMin
+	if rollupTime != 0 {
+		newPart.rollupBuckets = int(math.Ceil(float64(partitionInterval) / float64(rollupTime)))
 	}
 
 	return &newPart
 }
 
 type PartitionManager struct {
-	mtx           sync.RWMutex
-	path          string
-	cfg           *config.DBPartConfig
-	headPartition *DBPartition
-	cyclic        bool
-	ignoreWrap    bool
+	mtx                      sync.RWMutex
+	path                     string
+	cfg                      *config.Schema
+	headPartition            *DBPartition
+	partitions               []*DBPartition
+	cyclic                   bool
+	container                *v3io.Container
+	currentPartitionInterval int64 //TODO update on schema changes
 }
 
 func (p *PartitionManager) IsCyclic() bool {
 	return p.cyclic
 }
 
-func (p *PartitionManager) GetConfig() *config.DBPartConfig {
+func (p *PartitionManager) GetPartitions() []*DBPartition {
+	return p.partitions
+}
+
+func (p *PartitionManager) GetConfig() *config.Schema {
 	return p.cfg
 }
 
 func (p *PartitionManager) Init() error {
-
 	return nil
 }
 
-func (p *PartitionManager) TimeToPart(t int64) *DBPartition {
-
-	return p.headPartition // TODO: find the matching partition, if newer create one
+func (p *PartitionManager) TimeToPart(t int64) (*DBPartition, error) {
+	if p.headPartition == nil {
+		_, err := p.createNewPartition(p.currentPartitionInterval * (t / p.currentPartitionInterval))
+		return p.headPartition, err
+	} else {
+		if t >= p.headPartition.startTime {
+			if (t - p.headPartition.startTime) > p.currentPartitionInterval {
+				_, err := p.createNewPartition(p.headPartition.startTime + p.currentPartitionInterval)
+				if err != nil {
+					return nil, err
+				}
+				return p.TimeToPart(t)
+			} else {
+				return p.headPartition, nil
+			}
+		} else {
+			//iterate backwards, ignore last elem as it's the headPartition
+			for i := len(p.partitions) - 2; i >= 0; i-- {
+				if t > p.partitions[i].startTime {
+					return p.partitions[i], nil
+				}
+			}
+		}
+	}
+	return p.headPartition, nil
 }
 
-func (p *PartitionManager) PartsForRange(mint, maxt int64) []*DBPartition {
-
-	return []*DBPartition{p.headPartition}
+func (p *PartitionManager) createNewPartition(t int64) (*DBPartition, error) {
+	time := t & 0x7FFFFFFFFFFFFFF0
+	path := fmt.Sprintf("%s/%d/", p.path, time/1000)
```

Comment:
use `filepath.Join(...)`

Reply:
why? whats the advantage?

Reply:
Actually, we ought to be using `path.Join`. `filepath.Join` is system-dependent. As to why not using `Sprintf` instead, I would argue that there are minor advantages in terms of readability (can read left-to-right without looking for substitutions, purpose as a path is clear, no need to consider escape sequences).

Reply:
1. OS independent separator
2. it adds a Separator if necessary
3. Join calls Clean (https://golang.org/pkg/path/filepath/#Clean) on the result; in particular, all empty strings are ignored. On Windows, the result is a UNC path if and only if the first path element is a UNC path.

Reply:
Again, should be `path.Join`. `filepath.Join` is system-dependent... This is not a local path - should not depend on client OS.